lib/exename.cpp

```diff
@@ -16,58 +16,83 @@ limitations under the License.
 
 #include "exename.h"
 
-#include <limits.h>
-#include <stdlib.h>
-#include <string.h>
-#include <unistd.h>
-
-#include <sys/stat.h>
-#include <sys/types.h>
-
-#include "exceptions.h"
+#include <array>
+#include <cstring>
+#include <filesystem>
 
 namespace P4 {
 
-template <size_t N>
-static void convertToAbsPath(const char *const relPath, char (&output)[N]) {
-    output[0] = '\0';  // Default to the empty string, indicating failure.
+#if defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__) || defined(__NetBSD__)
+#include <unistd.h>
+
+#include <climits>
+#ifdef __APPLE__
+#include <mach-o/dyld.h>
+#endif
+#elif defined(_WIN32)
+#include <windows.h>
+#endif
 
-    char cwd[PATH_MAX];
-    if (!getcwd(cwd, sizeof(cwd))) return;
-    const size_t cwdLen = strlen(cwd);
-    if (cwdLen == 0) return;
-    const char *separator = cwd[cwdLen - 1] == '/' ? "" : "/";
+std::filesystem::path getExecutablePath(const std::filesystem::path &suggestedPath) {
+#if defined(__linux__) || defined(__FreeBSD__) || defined(__NetBSD__)
+    // Find the path of the executable.  We use a number of techniques that may fail or work on
```

Comment:
Maybe adopt https://github.com/gpakosz/whereami? There is also https://original.boost.org/doc/libs/1_84_0/doc/html/boost/dll/program_location_idm4410.html

Reply:
Figured we shouldn't depend on external dependencies for a utility like this. Particularly boost which we want to get rid of. Let me take a look at `whereami`

Reply:
Yes. Though there is bunch of domain knowledge here :(